### DISCUSSION TITLE
v3.0.3.8 Custom Extension Sub-folders

### DISCUSSION THREAD

**Discussion by Buddyboy-git** [opencart/opencart#11785] [Ideas]
My original suggestion is insufficient. So I went back to the drawing board and have a new code change suggestion. And funny enough, v4 does exactly what i am proposing here for v3.x.
The only problem is, I am still horrible at git, and the user voice feature doesn't work.
So I am proposing my code change here. And this allows for custom extensions to have subdirectories.
And since you guys are seasoned developers you will probably touch this up a bit.

Here is my new Admin/Controller/Startup/Permission
Code: [Select all](https://forum.opencart.com/viewtopic.php?t=229779#)
`
class ControllerStartupPermission extends Controller {
	public function index() {
		if (isset($this->request->get['route'])) {
			$route = '';
			$part = explode('/', $this->request->get['route']);

			**// New Code - Check if route is File or Function call.
			$i = 0;
			$arrayLength = count($part);
			$filename = DIR_APPLICATION . "controller/" . $this->request->get['route'] . ".php";
			if (!file_exists($filename)) {
				$arrayLength = $arrayLength - 1;
			}
			$permission_check_route = $part[0];;
			for ($i = 1; $i < $arrayLength; $i++) {
				$permission_check_route .= "/" . $part[$i] ;
			}
			// End New Code.**
``
			if (isset($part[0])) {
				$route .= $part[0];
			}
			if (isset($part[1])) {
				$route .= '/' . $part[1];
			}
			// If a 3rd part is found we need to check if its under one of the extension folders.
			$extension = array(
				'extension/advertise',
				'extension/dashboard',
				'extension/analytics',
				'extension/captcha',
				'extension/extension',
				'extension/feed',
				'extension/fraud',
				'extension/module',
				'extension/payment',
				'extension/shipping',
				'extension/theme',
				'extension/total',
				'extension/report'
			);
			if (isset($part[2]) && in_array($route, $extension)) {
				$route .= '/' . $part[2];
			}
			// We want to ingore some pages from having its permission checked.
			$ignore = array(
				'common/dashboard',
				'common/login',
				'common/logout',
				'common/forgotten',
				'common/reset',
				'error/not_found',
				'error/permission'
			);
			if (!in_array($route, $ignore) && !$this->user->hasPermission('access', **$permission_check_route** ) {
				return new Action('error/permission');
			}
		}
	}
}
`

**Comment by mhcwebdesign** [comment]
Quite confusing.

To start with: What problem exactly are you trying to solve?

**Reply by Buddyboy-git** [reply]
Currently v3 does not allow custom extensions to have sub-folders. 
This new code allows for custom extensions to have sub-folders. 
Currently Opencart throws a permission denied response because it does not search deep enough for the permission.
Example:  extension/extension/sale/order/info    will throw a permission denied!

**Reply by mhcwebdesign** [reply]
> Currently v3 does not allow custom extensions to have sub-folders. This new code allows for custom extensions to have sub-folders. Currently Opencart throws a permission denied response because it does not search deep enough for the permission. Example: extension/extension/sale/order/info will throw a permission denied!

Rightly so, it shouldn't allow sub-folders in admin/controller/extension/extension !

**Reply by Buddyboy-git** [reply]
Care to explain why?   Because, if it shouldn't be allowed, then the User Group Permissions shouldn't grab the sub-folders when they exist. 
I couldn't find any documentation on Extensions. And also what you are saying is that core developers are allowed to organize their code, but extension builders are not?

Thanks for any insight.
